rust/tools/authorization-logic/src/parsing/astconstructionvisitor.rs

```diff
@@ -136,13 +131,41 @@ fn construct_can_say_fact(ctx: &CanSayFactContext) -> AstFact {
     }
 }
 
+fn construct_binop(ctx: &BinopContextAll) -> AstBinop {
+    match ctx {
+        BinopContextAll::LtbinopContext(_) => AstBinop::LessThan,
+        BinopContextAll::GrbinopContext(_) => AstBinop::GreaterThan,
+        BinopContextAll::EqbinopContext(_) => AstBinop::Equals,
+        BinopContextAll::NebinopContext(_) => AstBinop::NotEquals,
+        BinopContextAll::LeqbinopContext(_) => AstBinop::LessOrEquals,
+        BinopContextAll::GeqbinopContext(_) => AstBinop::GreaterOrEquals,
+        _ => { panic!("construct_binop tried to build error"); }
+    }
+}
+
+fn construct_rvalue(ctx: &RvalueContextAll) -> AstRValue {
+    match ctx {
+        RvalueContextAll::FlatFactRvalueContext(ffctx) => {
+            AstRValue::FlatFactRValue { 
+                flat_fact: construct_flat_fact(&ffctx.flatFact().unwrap())
+            }
+        },
+        RvalueContextAll::BinopRvalueContext(bctx) => {
+            AstRValue::BinopRValue {
+                lnum: bctx.pred_arg(0).unwrap().get_text(),
+                binop: construct_binop(&bctx.binop().unwrap()),
+                rnum: bctx.pred_arg(1).unwrap().get_text()
```

Comment:
I'm seeing a lot of `unwrap`s here; what is being `unwrap`ed? Should it be handled in a more principled fashion rather than `panic`ing if it is not present?

Reply:
The antlr listener API that returns a `context` for the subexpressions of the parse rule returns an option type. Unwrapping is fine because this is a parse error if it tries to unwrap a `None` in this case, so panicking is right -- though if this were code that was going to live for a while giving better error messages rather than what the default panic gives you would be an improvement. (Though I think actually if there is a parse error it will cause a panic with a better descrption before it reaches this branch (or the other similar branches for which I give the default panics)).